src/main/kotlin/org/rust/bsp/toolwindow/BspProjectTreeStructure.kt

```diff
@@ -0,0 +1,185 @@
+package org.rust.bsp.toolwindow
+
+import ch.epfl.scala.bsp4j.BuildTarget
+import com.intellij.ide.projectView.PresentationData
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.ui.JBColor
+import com.intellij.ui.SimpleTextAttributes
+import com.intellij.ui.tree.AsyncTreeModel
+import com.intellij.ui.tree.StructureTreeModel
+import com.intellij.ui.treeStructure.CachingSimpleNode
+import com.intellij.ui.treeStructure.SimpleNode
+import com.intellij.ui.treeStructure.SimpleTreeStructure
+import org.rust.bsp.BspConstants
+import org.rust.bsp.service.BspConnectionService
+import org.rust.bsp.service.BspProjectViewService
+import org.rust.cargo.icons.CargoIcons
+import java.awt.Color
+
+class BspProjectTreeStructure(
+    tree: BspProjectsTree,
+    parentDisposable: Disposable,
+    val project: Project
+) : SimpleTreeStructure() {
+    private lateinit var bspConnection: BspConnectionService
+    private val treeModel = StructureTreeModel(this, parentDisposable)
+    private var root = BspSimpleNode.Root(mutableMapOf(), emptyList(), project)
+    private var targetById: MutableMap<String, BspNodeStatus> = mutableMapOf()
+
+    init {
+        tree.model = AsyncTreeModel(treeModel, parentDisposable)
+    }
+
+    override fun getRootElement(): Any = root
+
+    fun updateBspProjects(targets: List<BuildTarget>) {
+        val top = targets.map { it.id }.toMutableSet()
+        top.removeIf { it.uri == BspConstants.BSP_WORKSPACE_ROOT_URI }
+        val targetsWithoutRoot = targets.filter { it.id.uri != BspConstants.BSP_WORKSPACE_ROOT_URI }
+
+        for (t in targetsWithoutRoot) {
+            for (d in t.dependencies) {
+                top.remove(d)
```

Comment:
Can we use `top.removeAll(targetsWithoutRoot.flatmap { it.dependencies })` or something similar?